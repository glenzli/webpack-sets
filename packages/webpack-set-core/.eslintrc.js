module.exports = {
    extends: [
        require.resolve('@webpack-sets/eslint-rules'),
    ]
  };
  