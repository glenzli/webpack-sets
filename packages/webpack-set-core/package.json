{
  "name": "@g4iz/webpack-set-core",
  "version": "1.0.0",
  "description": "core functions for webpack set implement",
  "main": "dist/index.js",
  "types": "dist/types/index.d.ts",
  "scripts": {
    "lint": "eslint src & eslint test",
    "build": "rimraf -rf dist && compile",
    "start": "",
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/glenzli/webpack-sets.git"
  },
  "keywords": [
    "webpack"
  ],
  "author": "g4iz@outlook.com",
  "license": "MIT",
  "devDependencies": {
    "typescript": "~4.6.2",
    "@webpack-sets/eslint-rules": "~1.0.0",
    "@webpack-sets/compiler": "~1.0.0",
    "eslint": "~8.10.0",
    "@types/lodash-es": "~4.17.6",
    "@types/babel__core": "~7.1.18",
    "@types/node": "~17.0.21",
    "webpack": "~5.70.0",
    "rimraf": "~3.0.2",
    "jest": "~27.5.1",
    "@types/jest": "~27.4.1",
    "ts-jest": "~27.1.3"
  },
  "dependencies": {
    "lodash-es": "~4.17.21",
    "@babel/core": "~7.17.5"
  }
}
