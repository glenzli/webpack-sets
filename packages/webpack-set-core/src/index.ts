export * from './code_parser';
export * from './loader';
export * from './utility';
