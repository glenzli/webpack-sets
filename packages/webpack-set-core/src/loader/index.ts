export * from './loader_utility';
export * from './loader.interface';
