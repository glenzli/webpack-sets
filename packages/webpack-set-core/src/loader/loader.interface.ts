import { validateSchema } from 'webpack';

export type LoaderSchema = Parameters<typeof validateSchema>[0];
