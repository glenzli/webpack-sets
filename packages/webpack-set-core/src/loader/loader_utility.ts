import { Visitor, transformSync, TransformOptions } from '@babel/core';

export class LoaderUtility {
    /**
     * 仅通过 visitor 对代码进行变换，排除 babel 一些默认参数比如 babelrc 的影响
     */
    public static visitorTransformSync(
        visitors: Visitor[],
        source: string,
        inputSourceMap?: TransformOptions['inputSourceMap'],
        sourceMaps?: TransformOptions['sourceMaps'],
    ) {
        return transformSync(source, {
            plugins: visitors.map(visitor => ({ visitor })),
            configFile: false,
            babelrc: false,
            inputSourceMap,
            sourceMaps,
        });
    }
}
