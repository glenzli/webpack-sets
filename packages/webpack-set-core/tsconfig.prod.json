{
    "extends": "./tsconfig.dev.json",
    "compilerOptions": {
      "sourceMap": false
    }
  }