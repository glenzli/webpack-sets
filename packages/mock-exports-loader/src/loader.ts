import { uniq, forEach } from 'lodash';
import { LoaderDefinitionFunction, WebpackError } from 'webpack';
import { ModuleUtility, LoaderSchema, DEFUALT_ID } from '@g4iz/webpack-set-core';
import { parseExportIdsFrom, pasreExportIdsFromType } from './parse_export_ids';

export interface MockExportsLoaderOptions {
    /** 尝试从 type 文件中获取类型 */
    parseFromType?: boolean;
    /** 利用所有的 exports 合成为 default 导出 */
    generateDefault?: boolean;
    /** 自定义的 mock */
    createMocks?: () => Record<string, unknown>;
}

const LOADER = 'MockExportsLoader';

const schema: LoaderSchema = {
    type: 'object',
    additionalProperties: false,
    properties: {
        parseFromType: { type: 'boolean' },
        generateDefault: { type: 'boolean' },
        createMocks: { instancof: 'Function' },
    },
};

/**
 * 将指定模块编译为空mock，导出转换为空函数
 */
const loader: LoaderDefinitionFunction<MockExportsLoaderOptions> = function (this, source, sourceMap, meta) {
    const options = this.getOptions(schema);

    const exportIds = parseExportIdsFrom(source);
    const logger = this.getLogger(LOADER);
    logger.debug(`Types extracted from source '${this.resourcePath}'`, exportIds);

    if (options.parseFromType) {
        const typeEntry = ModuleUtility.resolveTypeFile(this.resourcePath);
        if (typeEntry) {
            // 尝试从类型中提取
            const typeExportIds = pasreExportIdsFromType(typeEntry);
            logger.debug(`Types extracted from type '${typeEntry}'`, typeExportIds);
            exportIds.push(...typeExportIds);
        } else {
            logger.warn(`Cannot find type file for '${this.resourcePath}'`);
        }
    }

    const mockIds = uniq(exportIds);

    if (mockIds.length > 0 || options.createMocks) {
        const mockSource = generateMockCode(mockIds, options);
        this.callback(null, mockSource, undefined, meta);
        return;
    }

    const warning = new WebpackError('[mock-exports-loader] Cannot find exports and no custom mock is provided.');
    warning.file = this.resourcePath;
    this._compilation?.warnings.push(warning);

    this.callback(null, source, sourceMap ?? undefined, meta);
};

function generateMockCode(mockIds: string[], options: MockExportsLoaderOptions) {
    const mocks: Record<string, string> = {};
    // 因为导出类型较难直观识别，导出均转换为空 function
    mockIds.forEach(fn => mocks[fn] = 'function() {}');

    const mockCodes = ['Object.defineProperty(exports, "__esModule", { value: true });'];

    // 应用传入的自定义mock
    const customMockFn = options.createMocks?.toString();
    if (customMockFn) {
        mockCodes.push(`Object.assign(exports, (${customMockFn})());`);
    }

    // 生成exports.x = xxx形式的代码
    forEach(mocks, (value, key) => {
        if (!options.generateDefault || key !== DEFUALT_ID) {
            mockCodes.push(`exports.${key} = ${value};`);
        }
    });

    // 合成默认导出
    if (options.generateDefault) {
        mockCodes.push(`exports.${DEFUALT_ID} = Object.assign({}, exports);`);
    }

    return mockCodes.join('\n');
}

export default loader;
