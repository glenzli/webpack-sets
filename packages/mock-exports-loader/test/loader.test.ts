import 'jest';
import fs from 'fs';
import loader from '../src/loader';
import { ModuleUtility } from '@g4iz/webpack-set-core';

describe('mockExportsLoader', () => {
    test('mock fs-extra', () => {
        const resourcePath = ModuleUtility.resolveModuleEntry('fs-extra');
        const callback = jest.fn();
        const logger = { debug: jest.fn(), warn: jest.fn() };
        loader.call(
            // @ts-ignore
            {
                getOptions: () => ({}),
                // @ts-ignore
                getLogger: () => logger,
                resourcePath,
                callback,
            },
            fs.readFileSync(resourcePath, { encoding: 'utf-8' }),
        );

        expect(callback.mock.calls[0][1]).toMatchSnapshot();
    });

    test('mock fs-extra with types and generate default', () => {
        const resourcePath = ModuleUtility.resolveModuleEntry('fs-extra');
        const callback = jest.fn();
        const logger = { debug: jest.fn(), warn: jest.fn() };
        loader.call(
            // @ts-ignore
            {
                getOptions: () => ({ parseFromType: true, generateDefault: true }),
                // @ts-ignore
                getLogger: () => logger,
                resourcePath,
                callback,
            },
            fs.readFileSync(resourcePath, { encoding: 'utf-8' }),
        );

        expect(callback.mock.calls[0][1]).toMatchSnapshot();
    });

    test('mock fs-extra with types, generate default and custom mocks', () => {
        const resourcePath = ModuleUtility.resolveModuleEntry('fs-extra');
        const callback = jest.fn();
        const logger = { debug: jest.fn(), warn: jest.fn() };
        loader.call(
            // @ts-ignore
            {
                getOptions: () => ({
                    parseFromType: true,
                    generateDefault: true,
                    createMocks: () => ({ CONST: 3, STR: 'test', FUNC: () => 1 }),
                }),
                // @ts-ignore
                getLogger: () => logger,
                resourcePath,
                callback,
            },
            fs.readFileSync(resourcePath, { encoding: 'utf-8' }),
        );

        expect(callback.mock.calls[0][1]).toMatchSnapshot();
    });
});
