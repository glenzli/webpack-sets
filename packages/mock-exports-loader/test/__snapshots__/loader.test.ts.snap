// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mockExportsLoader mock fs-extra 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.default = function() {};"
`;

exports[`mockExportsLoader mock fs-extra with types and generate default 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.copy = function() {};
exports.copySync = function() {};
exports.copyFile = function() {};
exports.move = function() {};
exports.moveSync = function() {};
exports.createFile = function() {};
exports.createFileSync = function() {};
exports.createSymlink = function() {};
exports.createSymlinkSync = function() {};
exports.ensureDir = function() {};
exports.ensureDirSync = function() {};
exports.mkdirs = function() {};
exports.mkdirp = function() {};
exports.mkdirsSync = function() {};
exports.mkdirpSync = function() {};
exports.outputFile = function() {};
exports.outputFileSync = function() {};
exports.readJson = function() {};
exports.readJSON = function() {};
exports.readJsonSync = function() {};
exports.readJSONSync = function() {};
exports.remove = function() {};
exports.removeSync = function() {};
exports.outputJSON = function() {};
exports.outputJson = function() {};
exports.outputJsonSync = function() {};
exports.outputJSONSync = function() {};
exports.writeJSON = function() {};
exports.writeJson = function() {};
exports.writeJsonSync = function() {};
exports.writeJSONSync = function() {};
exports.ensureFile = function() {};
exports.ensureFileSync = function() {};
exports.ensureLink = function() {};
exports.createLink = function() {};
exports.ensureLinkSync = function() {};
exports.createLinkSync = function() {};
exports.ensureSymlink = function() {};
exports.ensureSymlinkSync = function() {};
exports.emptyDir = function() {};
exports.emptydir = function() {};
exports.emptyDirSync = function() {};
exports.emptydirSync = function() {};
exports.pathExists = function() {};
exports.pathExistsSync = function() {};
exports.access = function() {};
exports.appendFile = function() {};
exports.chmod = function() {};
exports.chown = function() {};
exports.close = function() {};
exports.fchmod = function() {};
exports.fchown = function() {};
exports.fdatasync = function() {};
exports.fstat = function() {};
exports.fsync = function() {};
exports.ftruncate = function() {};
exports.futimes = function() {};
exports.lchown = function() {};
exports.link = function() {};
exports.lstat = function() {};
exports.mkdir = function() {};
exports.mkdirSync = function() {};
exports.open = function() {};
exports.opendir = function() {};
exports.read = function() {};
exports.readFile = function() {};
exports.readdir = function() {};
exports.readlink = function() {};
exports.realpath = function() {};
exports.namespace = function() {};
exports.rename = function() {};
exports.rm = function() {};
exports.rmdir = function() {};
exports.stat = function() {};
exports.symlink = function() {};
exports.truncate = function() {};
exports.unlink = function() {};
exports.utimes = function() {};
exports.write = function() {};
exports.writeFile = function() {};
exports.writev = function() {};
exports.mkdtemp = function() {};
exports.default = Object.assign({}, exports);"
`;

exports[`mockExportsLoader mock fs-extra with types, generate default and custom mocks 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", { value: true });
Object.assign(exports, (function () { return ({ CONST: 3, STR: 'test', FUNC: function () { return 1; } }); })());
exports.copy = function() {};
exports.copySync = function() {};
exports.copyFile = function() {};
exports.move = function() {};
exports.moveSync = function() {};
exports.createFile = function() {};
exports.createFileSync = function() {};
exports.createSymlink = function() {};
exports.createSymlinkSync = function() {};
exports.ensureDir = function() {};
exports.ensureDirSync = function() {};
exports.mkdirs = function() {};
exports.mkdirp = function() {};
exports.mkdirsSync = function() {};
exports.mkdirpSync = function() {};
exports.outputFile = function() {};
exports.outputFileSync = function() {};
exports.readJson = function() {};
exports.readJSON = function() {};
exports.readJsonSync = function() {};
exports.readJSONSync = function() {};
exports.remove = function() {};
exports.removeSync = function() {};
exports.outputJSON = function() {};
exports.outputJson = function() {};
exports.outputJsonSync = function() {};
exports.outputJSONSync = function() {};
exports.writeJSON = function() {};
exports.writeJson = function() {};
exports.writeJsonSync = function() {};
exports.writeJSONSync = function() {};
exports.ensureFile = function() {};
exports.ensureFileSync = function() {};
exports.ensureLink = function() {};
exports.createLink = function() {};
exports.ensureLinkSync = function() {};
exports.createLinkSync = function() {};
exports.ensureSymlink = function() {};
exports.ensureSymlinkSync = function() {};
exports.emptyDir = function() {};
exports.emptydir = function() {};
exports.emptyDirSync = function() {};
exports.emptydirSync = function() {};
exports.pathExists = function() {};
exports.pathExistsSync = function() {};
exports.access = function() {};
exports.appendFile = function() {};
exports.chmod = function() {};
exports.chown = function() {};
exports.close = function() {};
exports.fchmod = function() {};
exports.fchown = function() {};
exports.fdatasync = function() {};
exports.fstat = function() {};
exports.fsync = function() {};
exports.ftruncate = function() {};
exports.futimes = function() {};
exports.lchown = function() {};
exports.link = function() {};
exports.lstat = function() {};
exports.mkdir = function() {};
exports.mkdirSync = function() {};
exports.open = function() {};
exports.opendir = function() {};
exports.read = function() {};
exports.readFile = function() {};
exports.readdir = function() {};
exports.readlink = function() {};
exports.realpath = function() {};
exports.namespace = function() {};
exports.rename = function() {};
exports.rm = function() {};
exports.rmdir = function() {};
exports.stat = function() {};
exports.symlink = function() {};
exports.truncate = function() {};
exports.unlink = function() {};
exports.utimes = function() {};
exports.write = function() {};
exports.writeFile = function() {};
exports.writev = function() {};
exports.mkdtemp = function() {};
exports.default = Object.assign({}, exports);"
`;
